homeassistant/components/notify/xmpp.py

```diff
@@ -93,29 +118,217 @@ def __init__(self):
+
+            try:
+                # Uploading with XEP_0363
+                _LOGGER.debug("Timeout set to %ss", timeout)
+                url = await self.upload_file(timeout=timeout)
+                if url is None:
+                    raise FileUploadError("Could not upload file")
+            except (IqError, IqTimeout, XMPPError) as ex:
+                _LOGGER.error("Upload error, could not send message %s", ex)
+            except NotConnectedError as ex:
+                _LOGGER.error("Connection error %s", ex)
+            except FileTooBig as ex:
+                _LOGGER.error(
+                    "File too big for server, could not upload file %s", ex)
+            except UploadServiceNotFound as ex:
+                _LOGGER.error("UploadServiceNotFound: "
+                              " could not upload file %s", ex)
+            except FileUploadError as ex:
+                _LOGGER.error("FileUploadError, could not upload file %s", ex)
+            except requests.exceptions.SSLError as ex:
+                _LOGGER.error("Cannot establish SSL connection %s", ex)
+            except requests.exceptions.ConnectionError as ex:
+                _LOGGER.error("Cannot connect to server %s", ex)
+            except (FileNotFoundError,
+                    PermissionError,
+                    IsADirectoryError,
+                    TimeoutError) as ex:
+                _LOGGER.error("Error reading file %s", ex)
+            except FutTimeoutError as ex:
+                _LOGGER.error("The server did not respond in time, %s", ex)
             else:
-                self.send_message(mto=recipient, mbody=message, mtype='chat')
-            self.disconnect(wait=True)
+                _LOGGER.info("Upload success")
+
+                if room:
+                    _LOGGER.info("Sending file to %s", room)
+                    message = self.Message(sto=room, stype='groupchat')
+                else:
+                    _LOGGER.info("Sending file to %s", recipient)
+                    message = self.Message(sto=recipient, stype='chat')
+
+                message['body'] = url
+                # pylint: disable=invalid-sequence-index
+                message['oob']['url'] = url
+                try:
+                    message.send()
+                except (IqError, IqTimeout, XMPPError) as ex:
+                    _LOGGER.error("Could not send image message %s", ex)
+
+        async def upload_file(self, timeout=None):
+            """Upload file to Jabber server and return new URL.
+
+            upload a file with Jabber XEP_0363 from a remote URL or a local
+            file path and return a URL of that file.
+            """
+            if data.get(ATTR_URL_TEMPLATE):
+                _LOGGER.debug(
+                    "Got url template: %s", data.get(ATTR_URL_TEMPLATE))
+                templ = template_helper.Template(
+                    data.get(ATTR_URL_TEMPLATE), hass)
```

Comment:
We can use `data[ATTR_URL_TEMPLATE]` here because we already tested it in the `if`.

(And the same applies to the branches below)

Reply:
yes, will do that.